### PR TITLE
feat(dashboard): Add duplicate widget button in dashboard edit mode

### PR DESCRIPTION
**Pull request by edwardgou-sentry**
Adds a duplicate widget button to Dashboard Widgets in Edit mode
image

### CODE REVIEW

File: static/app/views/dashboardsV2/dashboard.tsx [hunk 1]
```diff
@@ -310,16 +311,17 @@ class Dashboard extends Component<Props, State> {
...
-    widgetCopy.tempId = undefined;
+    const widgetCopy = cloneDeep(
+      assignTempId({...widget, id: undefined, tempId: undefined})
+    );
 
     let nextList = [...dashboard.widgets];
     nextList.splice(index, 0, widgetCopy);
     nextList = generateWidgetsAfterCompaction(nextList);
 
+    onUpdate(nextList);
```
Comment (k-fish) [inline static/app/views/dashboardsV2/dashboard.tsx:324]: Why was this required?

**Review by k-fish — APPROVED** [review]
Looks good, question is an aside.

File: tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx [hunk 2]
```diff
@@ -4,6 +4,7 @@ import {
   mountWithTheme as rtlMountWithTheme,
```
Comment (priscilawebdev) [inline tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx:4]: We do not normally rename named exports. Should we just use `mountWithTheme`?

File: static/app/views/dashboardsV2/widgetCard/index.tsx [hunk 3]
```diff
@@ -94,6 +101,9 @@ class WidgetCard extends React.Component<Props> {
           <IconClick data-test-id="widget-edit" onClick={onEdit}>
             <IconEdit color="textColor" />
           </IconClick>
+          <IconClick data-test-id="widget-duplicate" onClick={onDuplicate}>
```
Comment (priscilawebdev) [inline static/app/views/dashboardsV2/widgetCard/index.tsx:104]: instead of a `data-test-id` attribute, should we use `aria-label` here?

File: tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx [hunk 4]
```diff
@@ -225,4 +223,45 @@ describe('Dashboards > Dashboard', () => {
...
+      expect(screen.getByTestId('widget-duplicate')).toBeInTheDocument();
+    });
+
+    it('duplicates the widget', async () => {
+      const dashboardWithOneWidget = {...mockDashboard, widgets};
+      const {rerender} = mount(dashboardWithOneWidget, organization);
+      userEvent.click(screen.getByTestId('widget-duplicate'));
+      rerender();
+      await waitFor(() => {
+        expect(screen.getAllByText('Test Discover Widget')).toHaveLength(2);
```
Comment (matejminar) [inline tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx:263]: Should we replace this with findAllByText?

File: tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx (hunk 2 above)
Comment (narsaynorath) [reply tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx:4]: This was my bad. We haven't fully converted tests to RTL so instead of making a separate file with RTL tests, the new ones I added were using this `rtlMountWithTheme`

Comment (edwardgou-sentry) [reply tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx]: Hey Priscila, some of our test files haven't been fully converted to RTL yet, so we've got files with a mix of enzyme and RTL. This naming is so we don't clash on importing `mountWithTheme` for now. We'll eventually update all the enzyme tests to RTL in the future and undo the rename.

File: static/app/views/dashboardsV2/widgetCard/index.tsx (hunk 3 above)
Comment (edwardgou-sentry) [reply static/app/views/dashboardsV2/widgetCard/index.tsx]: 👍 updated to aria-label!

**Review by matejminar — APPROVED** [review]

File: tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx [hunk 5]
```diff
@@ -225,4 +222,43 @@ describe('Dashboards > Dashboard', () => {
...
+          dashboard={dashboard}
+          organization={mockedOrg}
+          isEditing
+          onUpdate={newWidgets => {
+            widgets.splice(0, widgets.length, ...newWidgets);
+          }}
+          handleUpdateWidgetList={() => undefined}
+          handleAddCustomWidget={() => undefined}
+          onSetWidgetToBeUpdated={() => undefined}
+          router={initialData.router}
```
Comment (narsaynorath) [inline tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx:241]: What happens if `mockedOrg` is passed in? Can we make the assumption that `initialData.router` applies here as well?

File: tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx [hunk 6]
```diff
@@ -225,4 +222,43 @@ describe('Dashboards > Dashboard', () => {
...
+      const dashboardWithOneWidget = {...mockDashboard, widgets};
+      mount(dashboardWithOneWidget, organization);
+      expect(screen.getByLabelText('Duplicate Widget')).toBeInTheDocument();
+    });
+
+    it('duplicates the widget', async () => {
+      const dashboardWithOneWidget = {...mockDashboard, widgets};
+      const {rerender} = mount(dashboardWithOneWidget, organization);
+      userEvent.click(screen.getByLabelText('Duplicate Widget'));
+      rerender();
```
Comment (narsaynorath) [inline tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx:260]: Do we manually trigger a re-render here because there are a lot of updates that occur after `Duplicate Widget` is clicked? I'm wondering if we've tried wrapping the `userEvent.click` with an `act`

**Review by narsaynorath — APPROVED** [review]
This is a great change, thanks! I had two comments about the tests because it looks different from what I'm normally used to seeing, so I just wanted to learn why it's different.

EDIT: I've also tested it out locally 👍

File: tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx (hunk 6 above)
Comment (edwardgou-sentry) [reply tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx:260]: We're not testing the `detail.tsx` layer here, just `dashboard.tsx`, which means we need to provide our own logic to update the `dashboard` prop whenever `onUpdate` is called (such as clicking `Duplicate Widget` in this case. This function is typically provided by `detail.tsx`):

```
          onUpdate={newWidgets => {
            widgets.splice(0, widgets.length, ...newWidgets);
          }}
```

As far as I can tell, RTL won't pickup the prop change itself, so we need to call rerender to trigger the update. `act` also doesn't seem to solve this here

File: tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx (hunk 5 above)
Comment (edwardgou-sentry) [reply tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx:241]: Good point, probably not. I actually don't think I need the org as an arg so I've just removed the arg for now and kept the organization as `initialData.organization` to avoid this issue. Thanks!

File: tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx (hunk 6 above)
Comment (narsaynorath) [reply tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx:260]: Ah okay, this is neat. So you're defining `const widgets`, and the `onUpdate` inside `mount` updates that widget array and when we call rerender it uses the new list of widgets as props.

One thing I'd be concerned about is leaking state between tests since that `widgets` variable is shared across tests and is a `const` right? Wondering if we could localize the "widget state" inside the closure of `mount` so it's isolated on every instance of `mount`. e.g.

```js
const mount = dashboard => {
  const widgets = [...dashboard.widgets];
  // everything else
};
```

Comment (narsaynorath) [reply tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx:260]: And we would just be overriding the widgets we pass in through the dashboard a bit more explicitly since there's no more global `widgets` variable

Comment (edwardgou-sentry) [reply tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx:264]: Good point, I just added a before each to set the `widgets` so there shouldn't be any leaks now. The individual widgets don't get mutated within the tests so I think this should be fine for now!

**Review by priscilawebdev — APPROVED** [review]
looks good to me
